cpp/src/transform/row_bit_count.cu

```diff
@@ -0,0 +1,537 @@
+ *
+ * `depth` represents our depth in the source column hierarchy.
+ *
+ * "branches" within the spans can occur when we have lists inside of structs.
+ * consider a case where we are entering a struct<list, float> with a span of [4, 8].
+ * The internal list column will change that span to something else, say [5, 9].
+ * But when we finish processing the list column, the final float column wants to
+ * go back and use the original span [4, 8].
+ *
+ * [4, 8]  [5, 9]   [4, 8]
+ * struct< list<>   float>
+ *
+ * To accomplish this we maintain a stack of spans. Pushing the current span
+ * whenever we enter a branch, and popping a span whenever we leave a branch.
+ *
+ * `branch_depth_start` represents the branch depth as we reach a new column.
+ * if `branch_depth_start` is < the last branch depth we saw, we are returning
+ * from a branch and should pop off the stack.
+ *
+ * `branch_depth_end` represents the new branch depth caused by this column.
+ * if branch_depth_end > branch_depth_start, we are branching and need to
+ * push the current span on the stack.
+ *
+ */
+struct column_info {
+  size_type depth;
+  size_type branch_depth_start;
+  size_type branch_depth_end;
+};
+
+/**
+ * @brief Struct which contains hierarchy information precomputed on the host.
+ *
+ * If the input data contains only fixed-width types, this preprocess step
+ * produces the value `simple_per_row_size` which is a constant for every
+ * row in the output.  We can use this value and skip the more complicated
+ * processing for lists, structs and strings entirely if `complex_type_count`
+ * is 0.
+ *
+ */
+struct hierarchy_info {
+  hierarchy_info() : simple_per_row_size(0), complex_type_count(0), max_branch_depth(0) {}
+
+  // These two fields act as an optimization. If we find that the entire table
+  // is just fixed-width types, we do not need to do the more expensive kernel call that
+  // traverses the individual columns. So if complex_type_count is 0, we can just
+  // return a column where every row contains the value simple_per_row_size
+  size_type simple_per_row_size;  // in bits
+  size_type complex_type_count;
+
+  // max depth of span branches present in the hierarchy.
+  size_type max_branch_depth;
+};
+
+/**
+ * @brief Function which flattens the incoming column hierarchy into a vector
+ * of column_views and produces accompanying column_info and hierarchy_info
+ * metadata.
+ *
+ * @param begin: Beginning of a range of column views
+ * @param end: End of a range of column views
+ * @param out: (output) Flattened vector of output column_views
+ * @param info: (output) Additional per-output column_view metadata needed by the gpu
+ * @param h_info: (output) Information about the hierarchy
+ * @param cur_depth: Current absolute depth in the hierarchy
+ * @param cur_branch_depth: Current branch depth
+ * @param parent_index: Index into `out` representing our owning parent column
+ */
```

Comment:
I noticed that #7422 is also implementing a flatten-nested-columns function in `src/structs/utilities.hpp`
Is there some way to reuse that or for that PR to reuse this code?

Reply:
In-fact, there's a separate kind of flattening happening in parquet writer where all structs are being converted into single child structs. That one enables a simple iteration but obviously has a problem that it will cause the struct's null mask to be considered multiple times.

Three developers, three different approaches. Should we sit together and develop a common one?

Reply:
This particular implementation is different than what's going on in 7422.  In particular:
- This only flattens the "root" columns.  That is, if you have a strings column, the chars and offsets are not part of the output.  For lists, the offsets are not included but the child is. 
- This handles lists as well as structs.
- This produces a chunk of metadata alongside the vector of column_views.

The primary point of commonality is just the recursion through the hierarchy. I suppose we could provide a templated wrapper for that that takes a lambda where the user could provide whatever code they want.  But even that isn't quite what we want here because the conditions for recursion (in this case, _not_ all children) aren't the same.